src/result/result.py

```diff
@@ -61,6 +67,13 @@ def is_ok(self) -> Literal[True]:
     def is_err(self) -> Literal[False]:
         return False
 
+    def __bool__(self) -> bool:
+        """
+        Ok is considered valid only if its non-None
+        it's the same as a return of Err.ok()
+        """
+        return self._value is not None
```

Comment:
🤔 i think this is confusing. `None` values are perfectly fine as a result value.

a `__bool__` method has semantical problems. typically only empty containers and ‘false-ish’ values (`None`, `0`, `""`, etc.) evaluate as `False`, and everything else as `True`.

interpreting a `Result` as a boolean could also mean `Ok` vs `Err`, but this has type checking issues; see also https://github.com/rustedpy/result/issues/61

Reply:
Removed.
I just put it in when I was thinking about the Maybe construct. Seems to be an useful check, especially for python<3.10.
Anyway, it would be better to implement it as some kind of explicit `is_valid` method ... 
But I failed to make it work with the type checking. I tried to make a special Ok[None] type and fork in Ok.__new__, but the generics simply catch everything. According to few searches I made, there is no way to represent Type Exclusion (like Any - None).

Reply:
I made it work :)

```
def is_some(result: Result[Optional[T], E]) -> TypeGuard[Ok[T]]:
    """A typeguard to check if a result is an Ok and has non-None value.
    return result.is_ok() and result.ok() is not None
```

Now this can be done:
```
async def foo(x: int) -> Optional[int]: ...
async def bar(x: int) -> None: ...
async def test():
    x = await (Ok(1) >= foo)
    y = await (x >= bar)  ## type error
    if is_some(x):
        y = await (x >= bar)  ## ok types
```